### PR TITLE
residency hospital matching

### PR DESCRIPTION
**Pull request by atefeh-asayesh**
residencymatch.dml
residencyMatchMain.dml
BuiltinResidencyMatchTest.java
Builtins.java

### CODE REVIEW

**Review by atefeh-asayesh — COMMENTED** [review]
Four files have been updated.

File: scripts/builtin/residencyMatchMain.dml [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+  return (Matrix[Double] ResidencyMatch)
+{
+#TODO PLEASE PAY ATTENTION
+# in this step we consider that  Residents Matrix is ORDERED   !!!!
+# in this step we consider that  Hospital  Matrix is UNORDERED !!!!
+
+print("STARTING Resident Hospital Match");
+#TODO set a finite number of maximum iterations so that the execution termiates after maximum iterations.
+
+print("\n")
```
Comment (Shafaq-Siddiqi) [inline scripts/builtin/residencyMatchMain.dml:94]: in DML we follow two-space indentation. Please indent this code block.

File: scripts/builtin/residencyMatchMain.dml [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+# 1.0,0.0,0.0
+#
+# Resident 1 has matched with Hospital 3 (since [1,3] is non-zero) at a preference level of 3.0.
+# Resident 2 has matched with Hospital 1 (since [2,1] is non-zero) at a preference level of 1.0.
+# Resident 3 has matched with Hospital 2 (since [3,2] is non-zero) at a preference level of 2.0.
+# --------------------------------------------------------------------------------------------
+m_residencyMatchMain = function(Matrix[Double] R, Matrix[Double] H,Matrix[Double] C )
+  return (Matrix[Double] ResidencyMatch)
+{
+#TODO PLEASE PAY ATTENTION
```
Comment (Shafaq-Siddiqi) [inline scripts/builtin/residencyMatchMain.dml:87]: Please explain this TODO

File: scripts/builtin/residencyMatchMain.dml [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+
+print("STARTING Resident Hospital Match");
+#TODO set a finite number of maximum iterations so that the execution termiates after maximum iterations.
+
+print("\n")
+print("STARTING RESIDENCY MATCH ALGORITHM");
+print("READING R  as residents AND H as Hospitals and also C as capacity...");
+
+m = nrow(R)
+n = ncol (R)
```
Comment (Shafaq-Siddiqi) [inline scripts/builtin/residencyMatchMain.dml:99]: remove the space and make it ncol(R)

File: scripts/builtin/residencyMatchMain.dml [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+Capacityrows = nrow(C)
```
Comment (Shafaq-Siddiqi) [inline scripts/builtin/residencyMatchMain.dml:100]: just to follow the consistent naming conventions please start the variable name with a lower-case.

File: scripts/builtin/residencyMatchMain.dml [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+  for(i in 1:m) {
+
+    while (as.scalar (StarM[i]) == 1)  {
+      SecondIndex = as.scalar (proposer_pointers[i])
+      HIndex[i] = as.scalar (R[i,SecondIndex])
+      prev_Residents_vector = rowMaxs(HospitalMatch) # we consider if the preference value is 1 it means it is the first selection of that hospital, the minimum value means most preference.
+      prevIndex_Residents_vector =  rowIndexMax(HospitalMatch)
+      if (as.scalar(HIndex[i]) != 0 ){
+
+        HosValue = as.scalar (H[as.scalar (HIndex[i]),i])
```
Comment (Shafaq-Siddiqi) [inline scripts/builtin/residencyMatchMain.dml:138]: remove unnecessary space after as.scalar

File: scripts/builtin/residencyMatchMain.dml [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+#TODO set a finite number of maximum iterations so that the execution termiates after maximum iterations.
+
+print("\n")
+print("STARTING RESIDENCY MATCH ALGORITHM");
+print("READING R  as residents AND H as Hospitals and also C as capacity...");
+
+m = nrow(R)
+n = ncol (R)
+Capacityrows = nrow(C)
+#######################################################################################################
```
Comment (Shafaq-Siddiqi) [inline scripts/builtin/residencyMatchMain.dml:101]: simplify the comments please do not use unnecessary "#" and "--" as it makes it difficult to read the code.

File: scripts/builtin/residencyMatchMain.dml [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+prevIndex_Residents_vector =  rowIndexMin(HospitalMatch)
+
+while(sum(StarM) > 0) {
+
+  for(i in 1:m) {
+
+    while (as.scalar (StarM[i]) == 1)  {
+      SecondIndex = as.scalar (proposer_pointers[i])
+      HIndex[i] = as.scalar (R[i,SecondIndex])
+      prev_Residents_vector = rowMaxs(HospitalMatch) # we consider if the preference value is 1 it means it is the first selection of that hospital, the minimum value means most preference.
```
Comment (Shafaq-Siddiqi) [inline scripts/builtin/residencyMatchMain.dml:134]: Please set a line-width/margin to 80 - 100 characters in your editor to break long lines.

File: src/test/scripts/functions/builtin/residencymatch.dml [hunk 8]
```diff
@@ -0,0 +1,10 @@
+# this dml file will run the  ResidencyMatchMain function in builtin functions  in scripts/builtin
+R = read($1)
+H = read($2)
+C = read($3)
+alg = $4
+if( alg == "residencymatch" )
```
Comment (Shafaq-Siddiqi) [inline src/test/scripts/functions/builtin/residencymatch.dml:6]: please remove this condition as you have only one algorithm to match.

File: src/test/scripts/functions/builtin/residencymatch.dml [hunk 9]
```diff
@@ -0,0 +1,10 @@
...
+[Out] = residencyMatchMain(R=R,H=H,C=C)
```
Comment (Shafaq-Siddiqi) [inline src/test/scripts/functions/builtin/residencymatch.dml:7]: please remove the square brackets as the function only return a single output, not a list.

File: src/test/java/org/apache/sysds/test/functions/builtin/BuiltinResidencyMatchTest.java [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+	private final static double eps = 0.0001;
+	private final static int rows = 3;
+	private final static int cols = 3;
+
+	@Override
+	public void setUp() {
+		addTestConfiguration(TEST_NAME,new TestConfiguration(TEST_CLASS_DIR, TEST_NAME,new String[]{"SM"}));
+	}
+	@Test
+	public void testResidencyMatchTest() {
```
Comment (Shafaq-Siddiqi) [inline src/test/java/org/apache/sysds/test/functions/builtin/BuiltinResidencyMatchTest.java:49]: Please add more tests with different input matrices as we have done in stable marriage test. Also, add tests for single node and spark execution contexts.

File: scripts/builtin/residencyMatchMain.dml (hunk 3 above)
Comment (atefeh-asayesh) [reply scripts/builtin/residencyMatchMain.dml:99]: Done.

File: src/test/java/org/apache/sysds/test/functions/builtin/BuiltinResidencyMatchTest.java (hunk 10 above)
Comment (atefeh-asayesh) [reply src/test/java/org/apache/sysds/test/functions/builtin/BuiltinResidencyMatchTest.java:49]: done

File: src/test/scripts/functions/builtin/residencymatch.dml (hunk 9 above)
Comment (atefeh-asayesh) [reply src/test/scripts/functions/builtin/residencymatch.dml:7]: done

File: src/test/scripts/functions/builtin/residencymatch.dml (hunk 8 above)
Comment (atefeh-asayesh) [reply src/test/scripts/functions/builtin/residencymatch.dml:6]: done

File: scripts/builtin/residencyMatchMain.dml (hunk 6 above)
Comment (atefeh-asayesh) [reply scripts/builtin/residencyMatchMain.dml:101]: done

File: scripts/builtin/residencyMatchMain.dml (hunk 4 above)
Comment (atefeh-asayesh) [reply scripts/builtin/residencyMatchMain.dml:100]: done

File: scripts/builtin/residencyMatchMain.dml (hunk 2 above)
Comment (atefeh-asayesh) [reply scripts/builtin/residencyMatchMain.dml:87]: done

File: src/test/java/org/apache/sysds/test/functions/builtin/BuiltinResidencyMatchTest.java (hunk 10 above)
Comment (Shafaq-Siddiqi) [reply src/test/java/org/apache/sysds/test/functions/builtin/BuiltinResidencyMatchTest.java:49]: Also, please do not delete the license files.

File: src/main/java/org/apache/sysds/common/Builtins.java [hunk 11]
```diff
@@ -32,31 +31,35 @@
...
  *
- * To add a new builtin script function, simply add the definition here
+ * * To add a new builtin script function, simply add the definition here
  * as well as a dml file in scripts/builtin with a matching name. On
  * building SystemDS, these scripts are packaged into the jar as well.
  */
+
 public enum Builtins {
 	//builtin functions
-	ABSTAIN("abstain", true),
```
Comment (Shafaq-Siddiqi) [inline src/main/java/org/apache/sysds/common/Builtins.java:41]: Do not delete the existing entries from the file.

**Pull request closed without merging**